indirect.h

```diff
@@ -140,31 +144,57 @@ class indirect {
     requires std::copy_constructible<T>
   {
     assert(other.p_ != nullptr);  // LCOV_EXCL_LINE
-    if constexpr (std::is_copy_assignable_v<T>) {
-      if (p_ == nullptr) {
-        T* mem = allocator_traits::allocate(alloc_, 1);
-        try {
-          allocator_traits::construct(alloc_, mem, *other);
-          p_ = mem;
-        } catch (...) {
-          allocator_traits::deallocate(alloc_, mem, 1);
-          throw;
+    if (this != &other) {
+      if constexpr (std::is_copy_assignable_v<T>) {
+        if (p_ == nullptr) {
+          T* mem = allocator_traits::allocate(alloc_, 1);
+          try {
+            allocator_traits::construct(alloc_, mem, *other);
+            p_ = mem;
+          } catch (...) {
+            allocator_traits::deallocate(alloc_, mem, 1);
+            throw;
+          }
+        } else {
+          *p_ = *other.p_;
         }
       } else {
-        *p_ = *other.p_;
+        indirect tmp(other);
+        this->swap(tmp);
+      }
+      if constexpr (allocator_traits::propagate_on_container_copy_assignment::
+                        value) {
+        alloc_ = other.alloc_;
       }
-    } else {
-      indirect tmp(other);
-      this->swap(tmp);
     }
     return *this;
   }
 
-  constexpr indirect& operator=(indirect&& other) noexcept {
+  constexpr indirect& operator=(indirect&& other) noexcept(
+      allocator_traits::propagate_on_container_move_assignment::value) {
```

Comment:
I belive this should also be conditional with or of allocator traits is_always_equal as is the case for std::vector swap https://eel.is/c++draft/vector.capacity
```suggestion
      allocator_traits::propagate_on_container_move_assignment::value || allocator_traits::is_always_equal::value) {
```

Reply:
Agreed. Can you make the change?

Reply:
Yes, just pushed up.  I'm working on the swap methods now

Reply:
Interestingly it looks like no major STL implementations support the case where propagate_on_container_move_assignment is false or allocator are not equal:

Microsoft: https://github.com/microsoft/STL/blob/f362f7d7bea87bd199c927a89718e929ed89b187/stl/inc/vector#L3452

libstdc++: https://github.com/gcc-mirror/gcc/blob/9a82cca91dee190d845ee1681440bc69a07c6d49/libstdc%2B%2B-v3/include/bits/stl_vector.h#L1591

libc++: https://github.com/llvm/llvm-project/blob/5d35273a32d239b7407338e13ed71b59174d6536/libcxx/include/vector#L2017